keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -347,6 +349,30 @@ correct order.
 [enables]:https://github.com/kubernetes/kubernetes/blob/220498b83af8b5cbf8c1c1a012b64c956d3ebf9b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L703
 [filter]:#updating-storageversion
 
+#### Agreements
+
+1. Storageversion updates will be triggered by CRD creates and updates
```

Comment:
IIUC you mean SV updates are triggered at CRD create/update watch events.

Reply:
How do we clean up SV if a CRD is deleted?

Reply:
I agree this we want to reconcile SV with CRDs.

We should be careful how we word this. I'd like to make sure we always reconcile SVs.  If we take a purely edge triggered approach then if the apiserver (in a single control plane node configuration) crashes, it could miss the event.  So we should perform reconciliation fully (i.e. take a level triggered approach).  It's quite possible that during a relist we'll find that a SV is not reconciled with CRDs, we should update it at that time..

Reply:
> IIUC you mean SV updates are triggered at CRD create/update watch events.

That's right. Updated to reflect the same.

Reply:
> How do we clean up SV if a CRD is deleted?

Yeah been thinking about that. We need to introduce a delete capability. There was probably no need for a delete SV behavior for built-in resources so I guess we dont have a deleteSV() in [staging/src/k8s.io/apiserver/pkg/storageversion](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/storageversion) as of now?

Reply:
> If we take a purely edge triggered approach then if the apiserver (in a single control plane node configuration) crashes, it could miss the event

When the server reboots it will reconcile all the CRDs, and write SVs accordingly

> I'd like to make sure we always reconcile SVs

One use case I can think of is by reconciling SVs, servers can fix SVs if there are additional malicious writers that modify SVs

Reply:
> We need to introduce a delete capability

Agreed. Let's make sure we have a cleanup strategy

Reply:
> How do we clean up SV if a CRD is deleted?

Why do we need to clean up SVs when a CRD is deleted?

If a CRD is deleted, it should trigger garbage collection, which means individual CRs (of that CRD) get deleted, which means we have no corresponding entry in etcd for individual objects (of that CRD).

Reply:
That's a fair point. For all SV objects we make their [owner ref to be the respective CRDs](https://github.com/kubernetes/kubernetes/pull/123999/files#diff-ad10c66dede77a3da8db2294a745acdef0fbdd4f7e7ac8d18b835572572943a3R332-R345). So once the CRDs go away, so should the SV objects. @roycaihw WDYT?